src/backend/src/modules/auth/auth.service.ts

```diff
@@ -0,0 +1,172 @@
+
+  private keycloakLogoutUri: string;
+
+  constructor(private readonly httpService: HttpService) {
+    this.keycloakAuthServerUri = configService.getValue(
+      'KEYCLOAK_AUTH_SERVER_URI',
+    );
+    this.keycloakResponseType = configService.getValue(
+      'KEYCLOAK_RESPONSE_TYPE',
+    );
+    this.keycloakScope = configService.getValue('KEYCLOAK_SCOPE');
+    this.keycloakRedirectUri = configService.getValue('KEYCLOAK_REDIRECT_URI');
+    this.keycloakClientId = configService.getValue('KEYCLOAK_CLIENT_ID');
+    this.keycloakClientSecret = configService.getValue(
+      'KEYCLOAK_CLIENT_SECRET',
+    );
+    this.keycloakUserInfoUri = configService.getValue('KEYCLOAK_USER_INFO_URI');
+    this.keycloakTokenUri = configService.getValue('KEYCLOAK_TOKEN_URI');
+    this.keycloakLogoutUri = configService.getValue('KEYCLOAK_LOGOUT_URI');
+  }
+
+  getUrlLogin(): any {
+    return {
+      url:
+        `${this.keycloakAuthServerUri}` +
+        `?client_id=${this.keycloakClientId}` +
+        `&response_type=${this.keycloakResponseType}` +
+        `&scope=${this.keycloakScope}` +
+        `&redirect_uri=${this.keycloakRedirectUri}`,
+    };
+  }
+
+  async getAccessToken(code: string): Promise<KeycloakToken> {
+    const params = {
+      grant_type: 'authorization_code',
+      client_id: this.keycloakClientId,
+      client_secret: this.keycloakClientSecret,
+      code: code,
+      redirect_uri: this.keycloakRedirectUri,
+    };
+
+    const data = await firstValueFrom(
+      this.httpService
+        .post(
+          this.keycloakTokenUri,
+          queryString.stringify(params),
+          this.getContentType(),
+        )
+        .pipe(
+          map(
+            (res: any) =>
+              new KeycloakToken(
+                res.data.access_token,
+                res.data.refresh_token,
+                res.data.expires_in,
+                res.data.refresh_expires_in,
+              ),
+          ),
+          catchError((e) => {
+            throw new HttpException(e.response.data, e.response.status);
+          }),
+        ),
+    );
+    return data;
+  }
+
+  async getUserInfo(accessToken: string): Promise<KeycloakUser> {
+    const params = {
+      headers: {
+        Authorization: `Bearer ${accessToken}`,
+      },
+    };
+    const data = await firstValueFrom(
+      this.httpService.get(this.keycloakUserInfoUri, params).pipe(
+        map((res: any) => new KeycloakUser(res.data.name)),
+        catchError((e) => {
+          throw new HttpException(e.response.data, e.response.status);
+        }),
+      ),
+    );
+
+    return data;
+  }
+
+  async refreshAccessToken(refresh_token: string): Promise<KeycloakToken> {
+    const params = {
+      grant_type: 'refresh_token',
+      client_id: this.keycloakClientId,
+      client_secret: this.keycloakClientSecret,
+      refresh_token: refresh_token,
+      redirect_uri: this.keycloakRedirectUri,
+    };
+
+    const data = await firstValueFrom(
```

Comment:
If we are just awaiting on the first value anyways, do we really need rxjs interference here? or can we just solve it by Promise itself? 🤔 🤔 🤔

Reply:
Previous we can do httpService.post().toPromise() but toPromise got deprecated
https://stackoverflow.com/questions/51910908/nestjs-async-httpservice-call